internal/garbagecollector/garbageCollector.go

```diff
@@ -0,0 +1,90 @@
+package garbagecollector
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/jinzhu/gorm"
+	clusterPkg "github.com/openshift/assisted-service/internal/cluster"
+	"github.com/openshift/assisted-service/internal/host"
+	"github.com/openshift/assisted-service/pkg/leader"
+	"github.com/openshift/assisted-service/pkg/s3wrapper"
+	"github.com/openshift/assisted-service/restapi/operations/installer"
+	"github.com/sirupsen/logrus"
+)
+
+type Config struct {
+	DeletedUnregisteredAfter time.Duration `envconfig:"DELETED_UNREGISTERED_AFTER" default:"168h"` // 7d
+	DeregisterInactiveAfter  time.Duration `envconfig:"DELETED_INACTIVE_AFTER" default:"720h"`     // 20d
+}
+
+type GarbageCollectors interface {
+	PermanentClustersDeletion(ctx context.Context, olderThan strfmt.DateTime, objectHandler s3wrapper.API) error
+	DeregisterClusterInternal(ctx context.Context, params installer.DeregisterClusterParams) error
+}
+
+func NewGarbageCollectors(
+	db *gorm.DB,
+	log logrus.FieldLogger,
+	hostApi host.API,
+	clusterApi clusterPkg.API,
+	objectHandler s3wrapper.API,
+	leaderElector leader.Leader,
+) *garbageCollector {
+	return &garbageCollector{
+		db:            db,
+		log:           log,
+		hostApi:       hostApi,
+		clusterApi:    clusterApi,
+		objectHandler: objectHandler,
+		leaderElector: leaderElector,
+	}
+}
+
+type garbageCollector struct {
+	Config
+	db            *gorm.DB
+	log           logrus.FieldLogger
+	hostApi       host.API
+	clusterApi    clusterPkg.API
+	objectHandler s3wrapper.API
+	leaderElector leader.Leader
+}
+
+func (g garbageCollector) DeregisterInactiveClusters() {
```

Comment:
what's the algorithm? Deregister  a cluster every hour? We have 2000 clusters there - when will it finish?

Reply:
every hour it will search for all inactive clusters and deregister them.
inactive means that the last update wasn't in the past 30 days.

Reply:
Yes, the first time might take a while

Reply:
It's done in the background so we don't really care how long it takes.

Reply:
but won't it crash the service? Is there a sleep between operations?

Reply:
We are not in the 70s it run in parallel to everything else

Reply:
maybe we need to limit the number of clusters that we deregister for every interval, lets sat a max of 10 clusters 
this way we can control the load and limit the damage in case of a problem

Reply:
I would say 100
every 8 seconds we refresh status for all the clusters and hosts in the system deleting 1000 cluster will not change much.